LangID update

rggdmonk opened the pull request:
Prepare pull request from: https://github.com/hplt-project/warc2text-runner/commit/be20b5406c287d430ed90e5d78ca1d0d82e5f580

Added 1) preprocessing from Laurie, 2) improved installation instructions, and 3) enhanced logging.

rggdmonk committed:
prepare pull request from https://github.com/hplt-project/warc2text-runner/commit/be20b5406c287d430ed90e5d78ca1d0d82e5f580

nvanva commented on src/warc2text_runner/two/fastertext_lid/proto_langid.py:
```diff
@@ -4,66 +4,116 @@
+import regex
+import ujson
+
+from src.warc2text_runner.two.fastertext_lid.basic_log import langid_logger
```
This does not work after pip install. The package name is warc2text_runner, not src. Please start with warc2text_runner.

rggdmonk replied:
Ok. I will fix it.

Keep in mind some limitations for src layout. See more https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/#

nvanva reviewed (changes requested):
The code looks nice except for the imports.

nvanva commented on src/warc2text_runner/two/fastertext_lid/proto_langid.py:
```diff
@@ -4,66 +4,116 @@
-                elif json_line["t"] == "":
-                    print(json.dumps({'lang': '_unk'}))
+                    self.logger.debug("Case: text is None.")
+                    print(ujson.dumps({"lang": ["_null"]}))
```
I would suggest using Python None here as the output for None input. It will simplify the code handling that: e.g. we will be able to write ```if json_line.get("lang"): ...``` instead of ```if json_line.get("lang")=="_null"```

ZJaume replied:
I agree, we should reduce the possible values of null.

rggdmonk replied:
Is this works?

Now:
```
{"lang":["langcode1", "langcode2", "langcode3"],"prob":[0.9975,0.0004,0.0004]} # normal case
{"lang":["_null"]} # get None from "t" field
{"lang":["_unk"]} # get empty string from "t" field
```

Will be:
```
{"lang":["langcode1", "langcode2", "langcode3"],"prob":[0.9975,0.0004,0.0004]}
{"lang":null}
{"lang":["_unk"]}
```

nvanva commented on src/warc2text_runner/two/fastertext_lid/proto_langid.py:
```diff
@@ -4,66 +4,116 @@
+
+        Example: "__label__eng_Latn" -> "eng_Latn"
+        """
+        return [label[9:] for label in prediction[0]]
```
May I suggest to make this prefix stripping more general and improve readability, this implementation doesn't account for FastText models with a little bit different label naming and also hard to read due to the explicit constant '9'. Maybe use ```label.replace('__label__','')``` instead?

rggdmonk replied:
In FastText, ```__label__``` is the default label prefix for classification, including language identification. See more: https://github.com/search?q=repo%3Afacebookresearch%2FfastText%20__label__&type=code

I will fix it.

nvanva commented on src/warc2text_runner/two/fastertext_lid/proto_langid.py:
```diff
@@ -4,66 +4,116 @@
+                        ujson.dumps(
+                            {
+                                "lang": self._postprocess_predicted_labels(prediction),
+                                "prob": self._postprocess_predicted_probabilities(prediction),
```
Maybe pass only ```prediction[0]``` and ```prediction[1]``` to the corresponding function since they don't need the whole tuple. I think this will improve readability a bit.

rggdmonk replied:
The `prediction[0]` and `prediction[1]` approach requires you to remember the functions of each post-process. Now, it simply retrieves output from FastText and performs operations.

nvanva reviewed (commented):
Minor issues, but easy to fix.

ZJaume commented on src/warc2text_runner/two/fastertext_lid/proto_langid.py:
```diff
@@ -4,66 +4,116 @@
+
+                elif len(json_line["t"]) == 0:
+                    self.logger.debug("Case: text is empty.")
+                    print(ujson.dumps({"lang": ["_unk"]}))
```
I would also use None here and leave "unk" (not _unk) for the cases where there is text and we do not know the language for any model-related reason (whatever confidence threshold we may use, for example).

rggdmonk committed:
refactor(langid): add updates for pull-request

* update for src layout https://github.com/hplt-project/warc2text-runner/pull/9#discussion_r1595560181
* https://github.com/hplt-project/warc2text-runner/commit/621ad65a1e4af2a6784ad09acaa6977145d0587a
* https://github.com/hplt-project/warc2text-runner/commit/9156dfb91db7c20fb583cdd37db7f554206c6d42

BREAKING CHANGE:
Before:
```
{"lang":["langcode1", "langcode2", "langcode3"],"prob":[0.9975,0.0004,0.0004]} # normal case
{"lang":["_null"]} # get None from "t" field
{"lang":["_unk"]} # get empty string from "t" field
```

After:
```
{"lang":["langcode1", "langcode2", "langcode3"],"prob":[0.9975,0.0004,0.0004]}
{"lang":null}
{"lang":["_unk"]}
```

nvanva commented:
Thank you for timely update!
